internal/ociwclayer/export.go

```diff
@@ -69,6 +71,27 @@ func writeTarFromLayer(r hcsshim.LayerReader, w io.Writer) error {
 				return err
 			}
 		} else {
+			fileIDInfo, err := r.LinkInfo()
+			if err != nil {
+				return err
+			}
+			if linkName, ok := linkRecords[fileIDInfo.FileID]; ok {
+				// We've seen this file before, by another name, so put a hardlink in the tar stream.
+				hdr := backuptar.BasicInfoHeader(name, 0, fileInfo)
+				hdr.Mode = 0644
+				hdr.Typeflag = tar.TypeLink
+				hdr.Linkname = linkName
+				if err := t.WriteHeader(hdr); err != nil {
+					return err
+				}
+				continue
+			}
+
+			// All subsequent names for this file will be hard-linked to this name
+			// Optimisation opportunity: Only store records for files with multiple links.
+			// e.g., go-winio could expose FileStandardInfo to enable this.
```

Comment:
I've opened https://github.com/microsoft/go-winio/pull/185 to get access to this data.